tutorials/en/vectara-advanced-app-tutorial-showcase-the-creation-of-vectara-app-in-legal-or-customer-support-use-case.mdx

```diff
@@ -0,0 +1,671 @@
+```
+
+The first step is importing the necessary libraries. Libraries like `requests` and `OAuth2Session` from `authlib` are fundamental for handling HTTP requests and OAuth2 authentication respectively, which are critical when communicating with Vectara's APIs. We also import `dotenv` to load environment variables from a `.env` file, ensuring a secure and organized way to handle configurations.
+
+### **2. Preparing the Environment** 🌿
+
+```python
+# Load environment variables from .env file
+load_dotenv()
+```
+By invoking `load_dotenv()`, we ensure that our script has access to crucial environment variables stored in a `.env` file. This not only enhances security but also promotes code reusability across different environments.
+
+### **3. Unveiling the Indexing Class** 🗂️
+
+```python
+class Indexing:
+    def __init__(self):
+        self.auth_url = os.getenv('AUTH_URL')
+        self.app_client_id = os.getenv('APP_CLIENT_ID')
+        self.app_client_secret = os.getenv('APP_CLIENT_SECRET')
+        self.jwt_token = self._get_jwt_token()
+        ...
+```
+The `Indexing` class is where the magic of data ingestion and indexing happens. Its methods are crafted to interact with Vectara's indexing API, laying down the tracks for our data to travel from our local environment to Vectara's corpus.
+
+#### **Securing Access with JWT Token** 🔐
+
+```python
+    def _get_jwt_token(self) -> str:
+        """Connect to the server and get a JWT token."""
+        token_endpoint = f"{self.auth_url}/oauth2/token"
+        session = OAuth2Session(self.app_client_id, self.app_client_secret, scope="")
+        token = session.fetch_token(token_endpoint, grant_type="client_credentials")
+        return token["access_token"]
+        ...
+```
+In `_get_jwt_token`, we initiate an OAuth2 session to obtain a JWT token, which is indispensable for authenticating our requests to Vectara's API.
+
+#### **Uploading Documents to the Corpus** 📤
+
+```python
+    def upload_file(self, customer_id: int, corpus_id: int, idx_address: str, uploaded_file, file_title: str) -> Tuple[requests.Response, bool]:
+            """Uploads a file to the corpus."""
+            # Determine the MIME type based on the file extension
+            extension_to_mime_type = {
+                '.txt': 'text/plain',
+                '.pdf': 'application/pdf',
+                '.doc': 'application/msword',
+                '.docx': 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
+                # ... add more mappings as needed
+            }
+            file_extension = os.path.splitext(uploaded_file.name)[-1]
+            mime_type = extension_to_mime_type.get(file_extension, 'application/octet-stream')
+
+            post_headers = {
+                "Authorization": f"Bearer {self.jwt_token}"
+            }
+            
+            try:
+                file = uploaded_file.read()  
+                files = {"file": (file_title, file, mime_type)}
```

Comment:
I am not sure the mime_type is needed (including lines 291-297). Why are you adding this here? Generally in Vectara you just upload the binary data in the file and that's it. See for example https://docs.vectara.com/docs/getting-started-samples/rest_upload_file.py

Reply:
When I tried , I was getting error. I'll try it again

Reply:
strange. okay more important to make it work. 
See vectara-ingest for example on how it works in another location: https://github.com/vectara/vectara-ingest/blob/main/core/indexer.py#L113
There the third argument is not used at all, and note https://github.com/vectara/vectara-ingest/blob/main/core/indexer.py#L133 where the file is read as binary stream. What does "uploaded_file.read()" do in this tutorial? where is that shown?